File: FinancialBI_CSharp/Network/CRUD/AccountManager.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
class AccountManager
{

  //  Variables
  internal bool state_flag;

  private MultiIndexDictionary<UInt32, string, Account> m_accountList = new MultiIndexDictionary<UInt32, string, Account>();

  public event EventHandler ObjectInitialized;

  public event EventHandler Read;

  private bool status;

  private Account attributes;

  public event EventHandler CreationEvent;

  private bool status;

  private Int32 id;

  public event EventHandler UpdateEvent;

  private ErrorMessage status;

  private Int32 id;

  public event EventHandler DeleteEvent;

  private bool status;

  private UInt32 id;

  public event EventHandler UpdateListEvent;

  private bool status;

  private List<Tuple<byte, bool, string>> updateResults;

  internal AccountManager()
  {
    NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_READ_ACCOUNT_ANSWER, &SMSG_READ_ACCOUNT_ANSWER));
    NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_DELETE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_DELETE_ACCOUNT_ANSWER));
    NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_LIST_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_LIST_ACCOUNT_ANSWER));
    state_flag = false;
  }

  internal void CMSG_LIST_ACCOUNT()
  {
    ByteBuffer packet = new ByteBuffer(((UShort)(ClientMessage.CMSG_LIST_ACCOUNT)));
    packet.Release();
    NetworkManager.GetInstance().Send(packet);
  }

  private void SMSG_LIST_ACCOUNT_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == 0))
    {
      object nb_accounts = packet.ReadInt32();
      for (Int32 i = 1; (i <= nb_accounts); i++)
      {
        object tmp_account = Account.BuildAccount(packet);
        m_accountList.Set(tmp_account.Id, tmp_account.Name, tmp_account);
      }

      state_flag = true;
      ObjectInitialized();
    }
    else
    {
      state_flag = false;
    }

  }

  internal void CMSG_CREATE_ACCOUNT(ref Account p_account)
  {
    NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_CREATE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_CREATE_ACCOUNT_ANSWER));
    ByteBuffer packet = new ByteBuffer(((UShort)(ClientMessage.CMSG_CREATE_ACCOUNT)));
    p_account.Dump(packet, false);
    packet.Release();
    NetworkManager.GetInstance().Send(packet);
  }

  private void SMSG_CREATE_ACCOUNT_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == 0))
    {
      CreationEvent(true, int.Parse(packet.ReadUint32()));
    }
    else
    {
      CreationEvent(false, null);
    }

    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_CREATE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_CREATE_ACCOUNT_ANSWER));
  }

  private void SMSG_READ_ACCOUNT_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == 0))
    {
      object l_account = Account.BuildAccount(packet);
      m_accountList.Set(l_account.Id, l_account.Name, l_account);
      Read(true, l_account);
    }
    else
    {
      Read(false, null);
    }

  }

  internal void CMSG_UPDATE_ACCOUNT(ref Account p_account)
  {
    NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_UPDATE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_UPDATE_ACCOUNT_ANSWER));
    ByteBuffer packet = new ByteBuffer(((UShort)(ClientMessage.CMSG_UPDATE_ACCOUNT)));
    p_account.Dump(packet, true);
    packet.Release();
    NetworkManager.GetInstance().Send(packet);
  }

  private void SMSG_UPDATE_ACCOUNT_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == ErrorMessage.SUCCESS))
    {
      UpdateEvent(packet.GetError(), int.Parse(packet.ReadUint32()));
    }
    else
    {
      UpdateEvent(packet.GetError(), null);
    }

    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_UPDATE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_UPDATE_ACCOUNT_ANSWER));
  }

  internal void CMSG_CRUD_ACCOUNT_LIST(ref Dictionary<Int32, CRUDAction> p_operations) {
        NetworkManager.GetInstance().SetCallback(ServerMessage.SMSG_UPDATE_ACCOUNT_LIST_ANSWER, new System.EventHandler(this.SMSG_UPDATE_ACCOUNT_LIST_ANSWER));
        ByteBuffer packet = new ByteBuffer(((UShort)(ClientMessage.CMSG_UPDATE_ACCOUNT_LIST)));
        packet.WriteUint32(p_operations.Count);
        foreach (op in p_operations) {
            object l_account = this.GetAccount(op.Key);
            if ((l_account == null)) {
                // TODO: Continue For... Warning!!! not translated
            }
            
            packet.WriteUint8(op.Value);
            if ((op.Value == CRUDAction.DELETE)) {
                packet.WriteInt32(op.Key);
            }
            else {
                l_account.Dump(packet, true);
            }
            
        }
        
        packet.Release();
        NetworkManager.GetInstance().Send(packet);
    }

  internal void SMSG_UPDATE_ACCOUNT_LIST_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == 0))
    {
      List<Tuple<byte, bool, string>> updatesStatus = new List<Tuple<byte, bool, string>>();
      for (Int32 i = 0; (i <= packet.ReadUint32()); i++)
      {
        CRUDAction action = packet.ReadUint8();
        if (((action == CRUDAction.DELETE)
                    || (action == CRUDAction.UPDATE)))
        {
          packet.ReadInt32();
        }

        //  ignore id
        updatesStatus.Add(new Tuple<byte, bool, string>(action, packet.ReadBool(), packet.ReadString()));
      }

      UpdateListEvent(true, updatesStatus);
    }
    else
    {
      UpdateListEvent(false, null);
    }

    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_UPDATE_ACCOUNT_LIST_ANSWER, new System.EventHandler(this.SMSG_UPDATE_ACCOUNT_LIST_ANSWER));
  }

  internal void CMSG_DELETE_ACCOUNT(ref UInt32 id)
  {
    ByteBuffer packet = new ByteBuffer(((UShort)(ClientMessage.CMSG_DELETE_ACCOUNT)));
    packet.WriteUint32(id);
    packet.Release();
    NetworkManager.GetInstance().Send(packet);
  }

  private void SMSG_DELETE_ACCOUNT_ANSWER(ByteBuffer packet)
  {
    if ((packet.GetError() == 0))
    {
      UInt32 id = packet.ReadInt32;
      m_accountList.Remove(id);
      DeleteEvent(true, id);
    }
    else
    {
      DeleteEvent(false, 0);
    }

  }

  public Account GetAccount(UInt32 p_id)
  {
    return m_accountList[p_id];
  }

  public Account GetAccount(Int32 p_id)
  {
    return m_accountList[System.UInt32.Parse(p_id)];
  }

  public Account GetAccount(string p_name)
  {
    return m_accountList[p_name];
  }

  internal List<Account> GetAccountsList(ref string LookupOption) {
        List<Account> tmp_list = new List<Account>();
        List<Account.FormulaTypes> selection = new List<Account.FormulaTypes>();
        switch (LookupOption) {
            case GlobalEnums.AccountsLookupOptions.LOOKUP_ALL:
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS);
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT);
                selection.Add(Account.FormulaTypes.FORMULA);
                selection.Add(Account.FormulaTypes.HARD_VALUE_INPUT);
                break;
            case GlobalEnums.AccountsLookupOptions.LOOKUP_INPUTS:
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT);
                selection.Add(Account.FormulaTypes.HARD_VALUE_INPUT);
                break;
            case GlobalEnums.AccountsLookupOptions.LOOKUP_OUTPUTS:
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS);
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT);
                selection.Add(Account.FormulaTypes.FORMULA);
                break;
            case GlobalEnums.AccountsLookupOptions.LOOKUP_TITLES:
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS);
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT);
                selection.Add(Account.FormulaTypes.FORMULA);
                break;
        }
        foreach (id in m_accountList.Keys) {
            object l_account = this.GetAccount(id);
            if ((l_account == null)) {
                // TODO: Continue For... Warning!!! not translated
            }
            
            if (selection.Contains(l_account.FormulaType)) {
                tmp_list.Add(l_account);
            }
            
        }
        
        return tmp_list;
    }

  internal MultiIndexDictionary<UInt32, string, Account> GetAccountsDictionary()
  {
    return m_accountList;
  }

  internal Int32 GetIdFromName(ref string name)
  {
    if ((m_accountList[name] == null))
    {
      return 0;
    }

    return m_accountList[name].Id;
  }

  internal void LoadAccountsTV(ref Windows.Forms.TreeView TV)
  {
    TreeViewsUtilities.LoadTreeview(TV, m_accountList);
  }

  internal void LoadAccountsTV(ref VIBlend.WinForms.Controls.vTreeView TV)
  {
    VTreeViewUtil.LoadTreeview(TV, m_accountList);
  }

  protected override void finalize()
  {
    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_READ_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_READ_ACCOUNT_ANSWER));
    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_DELETE_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_DELETE_ACCOUNT_ANSWER));
    NetworkManager.GetInstance().RemoveCallback(ServerMessage.SMSG_LIST_ACCOUNT_ANSWER, new System.EventHandler(this.SMSG_LIST_ACCOUNT_ANSWER));
    base.Finalize();
  }
}

File: PPS/Models/CRUDs/New CRUDs/AccountManager.vb
' Account2.vb
'
' CRUD for accounts table - relation with c++ server
'
'
'
'
' Author: Julien Monnereau
' Created: 16/07/2015
' Last modified: 10/09/2015


Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq
Imports CRUD

Friend Class AccountManager : Inherits NamedCRUDManager(Of NamedHierarchyCRUDEntity)

#Region "Init"

    Friend Sub New()

        CreateCMSG = ClientMessage.CMSG_CREATE_ACCOUNT
        ReadCMSG = ClientMessage.CMSG_READ_ACCOUNT
        UpdateCMSG = ClientMessage.CMSG_UPDATE_ACCOUNT
        UpdateListCMSG = ClientMessage.CMSG_CRUD_ACCOUNT_LIST
        DeleteCMSG = ClientMessage.CMSG_DELETE_ACCOUNT
        ListCMSG = ClientMessage.CMSG_LIST_ACCOUNT

        CreateSMSG = ServerMessage.SMSG_CREATE_ACCOUNT_ANSWER
        ReadSMSG = ServerMessage.SMSG_READ_ACCOUNT_ANSWER
        UpdateSMSG = ServerMessage.SMSG_UPDATE_ACCOUNT_ANSWER
        UpdateListSMSG = ServerMessage.SMSG_CRUD_ACCOUNT_LIST_ANSWER
        DeleteSMSG = ServerMessage.SMSG_DELETE_ACCOUNT_ANSWER
        ListSMSG = ServerMessage.SMSG_LIST_ACCOUNT_ANSWER

        Build = AddressOf Account.BuildAccount

        InitCallbacks()

    End Sub

#End Region

#Region "Mappings"

    Friend Function GetAccountsList(ByRef LookupOption As String) As List(Of Account)

        Dim tmp_list As New List(Of Account)
        Dim selection As New List(Of Account.FormulaTypes)
        Select Case LookupOption
            Case GlobalEnums.AccountsLookupOptions.LOOKUP_ALL
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS)
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT)
                selection.Add(Account.FormulaTypes.FORMULA)
                selection.Add(Account.FormulaTypes.HARD_VALUE_INPUT)

            Case GlobalEnums.AccountsLookupOptions.LOOKUP_INPUTS
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT)
                selection.Add(Account.FormulaTypes.HARD_VALUE_INPUT)

            Case GlobalEnums.AccountsLookupOptions.LOOKUP_OUTPUTS
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS)
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT)
                selection.Add(Account.FormulaTypes.FORMULA)

            Case GlobalEnums.AccountsLookupOptions.LOOKUP_TITLES
                selection.Add(Account.FormulaTypes.AGGREGATION_OF_SUB_ACCOUNTS)
                selection.Add(Account.FormulaTypes.FIRST_PERIOD_INPUT)
                selection.Add(Account.FormulaTypes.FORMULA)
        End Select

        For Each id In m_CRUDDic.Keys
            Dim l_account As Account = GetValue(id)

            If l_account Is Nothing Then Continue For
            If selection.Contains(l_account.FormulaType) Then
                tmp_list.Add(l_account)
            End If
        Next
        Return tmp_list

    End Function

#End Region

#Region "Utilities"

    Friend Sub LoadAccountsTV(ByRef p_treeview As Windows.Forms.TreeView)
        TreeViewsUtilities.LoadTreeview(p_treeview, m_CRUDDic)

    End Sub

    Friend Sub LoadAccountsTV(ByRef p_treeview As VIBlend.WinForms.Controls.vTreeView)
        VTreeViewUtil.LoadTreeview(p_treeview, m_CRUDDic)
    End Sub

    Friend Sub LoadAccountsTVIcons(ByRef p_treeview As VIBlend.WinForms.Controls.vTreeView)
        VTreeViewUtil.LoadTreeviewIcons(p_treeview, m_CRUDDic)
    End Sub

#End Region

End Class
